platform/security/handler/security-handler-pki/src/main/java/org/codice/ddf/security/handler/pki/OcspChecker.java

```diff
@@ -0,0 +1,284 @@
+import javax.ws.rs.core.Response;
+import org.apache.cxf.jaxrs.client.WebClient;
+import org.bouncycastle.asn1.x509.Certificate;
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cert.ocsp.BasicOCSPResp;
+import org.bouncycastle.cert.ocsp.CertificateID;
+import org.bouncycastle.cert.ocsp.OCSPException;
+import org.bouncycastle.cert.ocsp.OCSPReq;
+import org.bouncycastle.cert.ocsp.OCSPReqBuilder;
+import org.bouncycastle.cert.ocsp.OCSPResp;
+import org.bouncycastle.cert.ocsp.RevokedStatus;
+import org.bouncycastle.cert.ocsp.SingleResp;
+import org.bouncycastle.cert.ocsp.UnknownStatus;
+import org.bouncycastle.operator.DigestCalculator;
+import org.bouncycastle.operator.DigestCalculatorProvider;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.jcajce.JcaDigestCalculatorProviderBuilder;
+import org.codice.ddf.cxf.client.ClientFactoryFactory;
+import org.codice.ddf.cxf.client.SecureCxfClientFactory;
+import org.codice.ddf.system.alerts.NoticePriority;
+import org.codice.ddf.system.alerts.SystemNotice;
+import org.osgi.service.event.Event;
+import org.osgi.service.event.EventAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OcspChecker implements OcspService {
+  private static final Logger LOGGER = LoggerFactory.getLogger(OcspChecker.class);
+  private static final String HTTPS = "https://";
+
+  private final ClientFactoryFactory factory;
+  private final EventAdmin eventAdmin;
+
+  private boolean ocspEnabled;
+  private String ocspServerUrl;
+
+  public OcspChecker(ClientFactoryFactory factory, EventAdmin eventAdmin) {
+    this.factory = factory;
+    this.eventAdmin = eventAdmin;
+  }
+
+  /**
+   * Checks the whether the given {@param certs} are revoked or not against the OCSP server.
+   *
+   * @param certs - an array of certificates to verify
+   * @return true if the certificates are not revoked, false if any of them are revoked.
+   */
+  public boolean passesOcspChecker(X509Certificate[] certs) {
+    if (!ocspEnabled || ocspServerUrl == null) {
+      LOGGER.debug("OCSP check is not enabled. Skipping.");
+      return true;
+    }
+
+    if (!ocspServerUrl.startsWith(HTTPS)) {
+      postErrorEvent("The provided URL was not an HTTPS URL.");
+      return true;
+    }
+
+    for (X509Certificate cert : certs) {
+      try {
+        Certificate certificate = convertToBouncyCastleCert(cert);
+        OCSPReq ocspRequest = generateOcspRequest(certificate);
+        Response response = sendOcspRequest(ocspRequest);
+        OCSPResp ocspResponse = createOcspResponse(response);
+        boolean certIsRevoked = validateOCSPResponse(ocspResponse);
+        if (certIsRevoked) {
+          SecurityLogger.audit(
+              "Certificate {} has been revoked by the OCSP server {}.", cert, ocspServerUrl);
+          LOGGER.debug("Certificate has been revoked by the OCSP server.");
+        }
+
+        return certIsRevoked;
```

Comment:
❓ Should we be returning at the first valid cert or continue to check all of the certs?

Reply:
Good catch. 👍